src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLContext;
+import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.nio.conn.ssl.SSLIOSessionStrategy;
+import org.apache.kafka.common.network.Mode;
+import org.apache.kafka.common.security.ssl.SslFactory;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.ErrantRecordReporter;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.bulk.BackoffPolicy;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkProcessor;
+import org.elasticsearch.action.bulk.BulkProcessor.Listener;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.client.indices.CreateIndexRequest;
+import org.elasticsearch.client.indices.GetIndexRequest;
+import org.elasticsearch.client.indices.GetMappingsRequest;
+import org.elasticsearch.client.indices.GetMappingsResponse;
+import org.elasticsearch.client.indices.PutMappingRequest;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.common.unit.TimeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ElasticsearchClient {
+
+  private static final Logger log = LoggerFactory.getLogger(ElasticsearchClient.class);
+
+  private static final String RESOURCE_ALREADY_EXISTS_EXCEPTION =
+      "resource_already_exists_exception";
+  private static final String VERSION_CONFLICT_EXCEPTION = "version_conflict_engine_exception";
+  private static final Set<String> MALFORMED_DOC_ERRORS = new HashSet<>(
+      Arrays.asList(
+          "mapper_parsing_exception",
+          "illegal_argument_exception",
+          "action_request_validation_exception"
+      )
+  );
+
+  private AtomicReference<ConnectException> error;
+  private BulkProcessor bulkProcessor;
+  private ConcurrentHashMap<String, SinkRecord> recordMap;
+  private ElasticsearchSinkConnectorConfig config;
+  private ErrantRecordReporter reporter;
+  private RestHighLevelClient client;
+
+  public ElasticsearchClient(
+      ElasticsearchSinkConnectorConfig config,
+      ErrantRecordReporter reporter
+  ) {
+    int pos = 0;
```

Comment:
if this is too much boiler plate for you, you can also use `List#toArray` below (the typed version)

Reply:
not quite sure what you mean, could you provide an example?